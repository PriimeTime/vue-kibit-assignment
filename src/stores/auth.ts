import { defineStore } from "pinia";
import { ref, type Ref } from "vue";

export const useAuthStore = defineStore("auth", {
  state: () => ({
    user: ref(null) as Ref<User | null>,
  }),
  actions: {
    setUser(user: User | null) {
      this.user = user;
    },
    logout() {
      this.user = null;
    },
  },
});
