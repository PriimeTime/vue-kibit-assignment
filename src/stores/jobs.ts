import { ref, computed, type Ref } from "vue";
import { defineStore } from "pinia";
import type { S } from "vitest/dist/chunks/config.Cy0C388Z.js";

export const useJobsStore = defineStore("jobs", {
  state: () => ({
    loading: false, // TODO add this to a global loading hook ad create a global loading component
    searchTerm: ref("") as Ref<string>,
    jobs: [] as Job[],
  }),
  getters: {
    filteredJobs: (state) => (searchString: string) => {
      if (!searchString) {
        return state.jobs;
      }

      const lowercasedSearch = searchString.toLowerCase();
      return state.jobs.filter(
        (job) =>
          job.title.toLowerCase().includes(lowercasedSearch) ||
          job.description.toLowerCase().includes(lowercasedSearch) ||
          job.tags.some((tag) => tag.toLowerCase().includes(lowercasedSearch))
      );
    },
  },
});
