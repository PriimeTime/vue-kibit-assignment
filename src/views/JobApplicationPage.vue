<script setup lang="ts">
import DefaultLayout from "@/layouts/DefaultLayout.vue";
import BaseCard from "@/components/BaseCard.vue";
import HeaderText from "@/components/HeaderText.vue";
import { useJobsStore } from "@/stores/jobs";
import { useRoute } from "vue-router";
import BaseButton from "@/components/BaseButton.vue";
import BaseInput from "@/components/BaseInput.vue";

const route = useRoute();
const jobStore = useJobsStore();

const job = jobStore.jobs.filter((job) => job.id === route.params.id)[0];
</script>

<template>
  <DefaultLayout>
    <section class="w-full flex justify-center items-center mt-10">
      <BaseCard class="p-6 shadow-lg">
        <div v-if="job" class="space-y-4">
          <HeaderText class="text-2xl font-bold text-gray-800 text-center">
            Application as {{ job.title }}
          </HeaderText>
          <form>
            <div class="flex flex-col gap-4">
              <BaseInput placeholder="Name" />
              <BaseInput placeholder="Email" />
              <BaseInput
                type="textarea"
                placeholder="Tell us about yourself..."
              />
              <BaseButton class="w-full">Submit</BaseButton>
            </div>
          </form>
        </div>
        <div v-else class="text-center space-y-2">
          <HeaderText class="text-xl font-semibold text-gray-800">
            Job not found
          </HeaderText>
        </div>
      </BaseCard>
    </section>
  </DefaultLayout>
</template>
