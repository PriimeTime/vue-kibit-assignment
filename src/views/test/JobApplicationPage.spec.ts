import { describe, it, expect, vi } from "vitest";
import { mount } from "@vue/test-utils";
import JobApplicationPage from "@/views/JobApplicationPage.vue";
import { createTestingPinia } from "@pinia/testing";
import { createRouter, createWebHistory } from "vue-router";
import { useJobsStore } from "@/stores/jobs";
import { useAuthStore } from "@/stores/auth";

const mockRouter = createRouter({
  history: createWebHistory(),
  routes: [
    {
      path: "/application-success",
      component: { template: "<div>Success</div>" },
    },
    { path: "/", component: { template: "<div>Home</div>" } },
  ],
});

vi.mock("@/api/jobsService", () => ({
  bindJobToUser: vi.fn(),
}));

describe("JobApplicationPage", () => {
  it("renders without crashing", () => {
    const wrapper = mount(JobApplicationPage, {
      global: {
        plugins: [
          mockRouter,
          createTestingPinia({
            createSpy: vi.fn,
          }),
        ],
      },
    });
    expect(wrapper.exists()).toBe(true);
  });

  it("matches snapshot", () => {
    const wrapper = mount(JobApplicationPage, {
      global: {
        plugins: [
          mockRouter,
          createTestingPinia({
            createSpy: vi.fn,
          }),
        ],
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
