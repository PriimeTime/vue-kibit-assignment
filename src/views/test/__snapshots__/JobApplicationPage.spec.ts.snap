// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`JobApplicationPage > matches snapshot 1`] = `
"<header class="flex h-16 bg-teal-900">
  <div class="flex items-center w-2/3 text-xl md:text-3xl whitespace-nowrap text-white p-2 hover:cursor-pointer"> Job Ad Plaform </div>
  <div class="flex justify-end w-1/3 p-2"><button class="p-2 w-20 lg:mx-2 rounded-md text-white text-xs md:text-xl border-teal-900 border-2 bg-teal-900 flex justify-center items-center whitespace-nowrap">Home</button>
    <!--v-if--><button class="p-2 w-20 lg:mx-2 rounded-md text-white text-xs md:text-xl border-teal-900 border-2 bg-teal-900 flex justify-center items-center whitespace-nowrap">Login</button>
  </div>
</header>
<main>
  <section class="w-full flex justify-center items-center mt-10">
    <div class="max-w-[40rem] rounded-xl md:shadow-xl p-10 w-full p-6 md:shadow-lg">
      <div class="text-center space-y-2">
        <div class="text-3xl font-bold text-xl font-semibold text-gray-800"> Job not found </div>
      </div>
    </div>
  </section>
</main>"
`;
